Implement DDO typings for V4

mariacarmina opened the pull request:
Fixes #585  .

Changes proposed in this PR:

- create class for V4 asset;
- create class for V4 service;
- add tests.

mariacarmina commented on ocean_lib/assets/v4/asset.py:
```diff
@@ -0,0 +1,137 @@
+class V4Asset:
+    """Asset class to create, import, export, validate Asset/DDO objects for V4."""
+
+    def __init__(
```
I wanted to create a named tuple that includes all these required parameters and just pass that named tuple instead of this list. What do you think @calina-c, @DMats?

calina-c replied:
If the v3 asset has arguments, I'd rather keep these as arguments too. Anyone can use the `*args` syntax if they want to

mariacarmina commented on ocean_lib/services/v4/service.py:
```diff
@@ -0,0 +1,142 @@
+
+    def __init__(
+        self,
+        service_type: str,
```
The comment ^^ applies for this class as well.

calina-c commented on ocean_lib/assets/v4/asset.py:
```diff
@@ -0,0 +1,137 @@
+
+        self.other_values = values
+
+    def _build_sample_metadata(self) -> dict:
```
what is the purpose of this function? If it is for tests, please add it as a test sample and not in the main class.

mariacarmina replied:
I am about to remove it.

calina-c commented on ocean_lib/services/v4/service.py:
```diff
@@ -0,0 +1,142 @@
+        data_token: str,
+        files: str,
+        timeout: float,
+        other_values: Optional[Dict[str, Any]] = None,
```
do we still allow other_values? Either way, I'd reorder the args so that other_values comes last.

mariacarmina replied:
I was thinking to keep them for the compute dictionary. But now I think that is suitable to have them just for the service class right?

calina-c replied:
I'm not sure. I would discuss this with @alexcos20 or @trentmc, should we support extra keys in the DDO?

calina-c reviewed (changes requested):
A few questions/changes suggested.

mariacarmina commented:
I have also added some functions for add trusted algorithms publishers in each service class. In rest I have refactored the V4Asset and V4Service classes as @calina-c and @DMats proposed. I have created the tests in test_asset.py for the new structure.

calina-c commented on ocean_lib/assets/v4/asset.py:
```diff
@@ -0,0 +1,244 @@
+        chain_id = values.pop("chainId")
+        context = values.pop("@context")
+
+        if "metadata" in values:
```
I think you can remove the assignment part, just use these pops in the cls() call. Except for services, of course, which needs some cajoling before sending it to the constructor :)

calina-c commented on ocean_lib/assets/v4/asset.py:
```diff
@@ -0,0 +1,244 @@
+    def from_dict(cls, dictionary: dict) -> "V4Asset":
+        """Import a JSON dict into this Asset."""
+        values = copy.deepcopy(dictionary)
+        id_key = "id" if "id" in values else "_id"
```
I'm pretty sure this is something that's obsolete.

mariacarmina replied:
the `id_key`?

calina-c replied:
the "_id" key.

calina-c commented on ocean_lib/assets/v4/asset.py:
```diff
@@ -0,0 +1,244 @@
+        """
+
+        data = {
+            "@context": ["https://w3id.org/did/v1"],
```
only hardcode this if it is somehow missing. if there is a `@context` property on the object, please use that one. Otherwise fall back to this.

calina-c commented on ocean_lib/assets/v4/asset.py:
```diff
@@ -0,0 +1,244 @@
+            "chainId": self.chain_id,
+        }
+
+        for value in self.services:
```
can you please make this a list comprehension?

calina-c commented on ocean_lib/assets/v4/asset.py:
```diff
@@ -0,0 +1,244 @@
+                service = value.as_dictionary()
+                self.services.append(service)
+
+        self.services = list(filter(lambda s: isinstance(s, dict), self.services))
```
if the result of as_dictionary is always a dictionary, why are we filtering again?

mariacarmina replied:
I have tested with as_dictionary and services are V4Service objects and that's why I did this filter again here.

calina-c commented on ocean_lib/assets/v4/asset.py:
```diff
@@ -0,0 +1,244 @@
+            "event",
+            "stats",
+        ]
+        attrs = list(
```
I love this!

calina-c commented on ocean_lib/assets/v4/test/test_asset.py:
```diff
@@ -0,0 +1,218 @@
+    get_sample_v4_ddo_with_compute_service,
+)
+
+ENCRYPTED_FILES_URLS = "0x044736da6dae39889ff570c34540f24e5e084f4e5bd81eff3691b729c2dd1465ae8292fc721e9d4b1f10f56ce12036c9d149a4dab454b0795bd3ef8b7722c6001e0becdad5caeb2005859642284ef6a546c7ed76f8b350480691f0f6c6dfdda6c1e4d50ee90e83ce3cb3ca0a1a5a2544e10daa6637893f4276bb8d7301eb35306ece50f61ca34dcab550b48181ec81673953d4eaa4b5f19a45c0e9db4cd9729696f16dd05e0edb460623c843a263291ebe757c1eb3435bb529cc1 ...
```
- if using something like this, first please move it to some different file, where it does not compromise readability for an otherwise excellently-written class
- also make sure to leave an indication (a commeent) of whether this is an encryption of something, e.g. what should it decrypt to, or if it is just a placeholder etc.
- it looks a bit long to me. Is it a real encrypted files string taken from provider, or is it something else, more random? Maybe my eye is fooled by the GitHub interface but an encrypted files to be around 2-3 lines.

mariacarmina replied:
I retrieved it from the DDO structure.

calina-c commented on ocean_lib/services/v4/service.py:
```diff
@@ -0,0 +1,159 @@
+            raise IndexError
+
+        return cls(
+            service_id,
```
I would say use the pops directly here, as well.

calina-c commented on ocean_lib/services/v4/service.py:
```diff
@@ -0,0 +1,159 @@
+
+        :return: array of values
+        """
+        return dict()
```
can you please leave a comment that this was left in for uniformity with v3?

calina-c commented on ocean_lib/services/v4/service.py:
```diff
@@ -0,0 +1,159 @@
+        """
+        return dict()
+
+    def get_trusted_algos_v4(self) -> list:
```
see my earlier comments, remove the v4 suffix.

calina-c commented on ocean_lib/services/v4/service.py:
```diff
@@ -0,0 +1,159 @@
+        }
+
+        if self.type == "compute":
+            if "compute" in self.compute_values:
```
???

mariacarmina replied:
I verify if the type of the service is "compute" and after that if the "compute" key is already there. If it's already there I update the compute values, otherwise I'll add the `compute` key. I had some issues if I didn't use `update`, because `compute` key appeared twice like `compute: { compute: {`.

calina-c commented on ocean_lib/services/v4/service.py:
```diff
@@ -0,0 +1,159 @@
+
+            values[key] = value
+
+        if self._values:
```
do we still have _values?

mariacarmina replied:
No, I will delete it

calina-c commented on tests/resources/ddo_helpers.py:
```diff
@@ -33,16 +33,40 @@ def get_metadata() -> dict:
+    with open(path, "r") as file_handle:
+        ddo = file_handle.read()
+    ddo_dict = json.loads(ddo)
+    if key not in ddo_dict.keys():
```
no need for the if statement. I think it's `ddo_dict.pop(key, None)`

mariacarmina commented on ocean_lib/assets/asset_resolver.py:
```diff
@@ -3,47 +3,36 @@
 
     if ddo:
-        return V3Asset(dictionary=ddo.as_dictionary())
+        ddo_dict = ddo.as_dictionary()
```
For point 1, as far as I saw, `get_asset_ddo` is a get request and it was part of V3. Should we keep it or?
For point 2, I agree.

calina-c replied:
my point is we are getting an object, then we are converting it into a dict and back into an object. Why are we not returning the ddo directly?

calina-c commented on ocean_lib/assets/trusted_algorithms.py:
```diff
@@ -26,24 +28,41 @@ def generate_trusted_algo_dict(
-            )
-        ),
-    }
+    if ddo.version.startswith("4."):
```
you can add these as typings in the asset, e.g. generate_trusted_algos. This "if" can be removed.

calina-c commented on ocean_lib/assets/trusted_algorithms.py:
```diff
@@ -63,141 +82,110 @@ def create_publisher_trusted_algorithms(
-        compute_service.attributes["main"]["privacy"] == privacy_values
-    ), "New trusted algorithm was not added. Failed when updating the privacy key. "
-    return trusted_algos
+    if asset.version.startswith("4."):
```
the content of these blocks is identical. The point was to remove the "if/else" and you can now do that.

calina-c commented on ocean_lib/assets/trusted_algorithms.py:
```diff
@@ -63,141 +82,110 @@ def create_publisher_trusted_algorithms(
-        compute_service.attributes["main"]["privacy"] == privacy_values
-    ), "New trusted algorithm was not added. Failed when updating the privacy key. "
-    return trusted_algo_publishers
+    if asset.version.startswith("4."):
```
same here, please remove the if/else. The content of both branches is the same.

calina-c commented on ocean_lib/assets/trusted_algorithms.py:
```diff
@@ -63,141 +82,110 @@ def create_publisher_trusted_algorithms(
+        compute_service
+    ), "Cannot add trusted algorithm to this asset because it has no compute service."
+
+    if asset.version.startswith("4."):
```
here as well.

calina-c commented on ocean_lib/assets/trusted_algorithms.py:
```diff
@@ -63,141 +82,110 @@ def create_publisher_trusted_algorithms(
-    if not trusted_algorithm_publishers:
-        raise ValueError(
-            f"Publisher {publisher_address} is not in trusted algorith publishers of this asset."
+    if asset.version.startswith("4."):
```
and here

calina-c commented on ocean_lib/assets/v4/asset.py:
```diff
@@ -0,0 +1,264 @@
+        values = copy.deepcopy(dictionary)
+        id_key = "id" if "id" in values else "_id"
+        if "services" in values:
+            services = [V4Service.from_dict(value) for value in values.pop("services")]
```
or a one-liner: services = [] if "services" not in values else [V4Service.from_dict(value) for value in values.pop("services")]

calina-c commented on ocean_lib/assets/v4/asset.py:
```diff
@@ -0,0 +1,264 @@
+        assert service is not None, "this asset does not have a compute service."
+
+        trusted_algorithms = trusted_algorithms if trusted_algorithms else []
+        if trusted_algorithms:
```
no need for this if. If the list is empty, "for" will iterate an empty list and not do anything.

calina-c reviewed (changes requested):
just one more unresolved issue

calina-c commented on ocean_lib/assets/asset_resolver.py:
```diff
@@ -31,8 +31,4 @@ def resolve_asset(did: str, metadata_cache_uri: str) -> Union[V3Asset, V4Asset]:
-            return V4Asset.from_json(ddo_dict)
-        else:
-            return V3Asset(dictionary=ddo_dict)
+        return ddo
```
I think you can return ddo directly, no need for the if. Otherwise please consider this approved.
